src/Libraries/Microsoft.Extensions.Caching.Hybrid/Internal/HybridCacheEventSource.cs

```diff
@@ -0,0 +1,192 @@
+
+    private const int EventIdLocalCacheHit = 1;
+    private const int EventIdLocalCacheMiss = 2;
+    private const int EventIdDistributedCacheGet = 3;
+    private const int EventIdDistributedCacheHit = 4;
+    private const int EventIdDistributedCacheMiss = 5;
+    private const int EventIdDistributedCacheFailed = 6;
+    private const int EventIdBackendExecuteStart = 7;
+    private const int EventIdBackendExecuteComplete = 8;
+    private const int EventIdBackendExecuteFailed = 9;
+    private const int EventIdLocalCacheWrite = 10;
+    private const int EventIdDistributedCacheWrite = 11;
+    private const int EventIdStampedeJoin = 12;
+
+    // fast local counters
+    private long _totalLocalCacheHit;
+    private long _totalLocalCacheMiss;
+    private long _totalDistributedCacheHit;
+    private long _totalDistributedCacheMiss;
+    private long _totalBackendExecute;
+    private long _currentBackendExecute;
+    private long _currentDistributedFetch;
+    private long _totalLocalCacheWrite;
+    private long _totalDistributedCacheWrite;
+    private long _totalStampedeJoin;
+
+#if !(NETSTANDARD2_0 || NET462)
+    // full Counter infrastructure
+    private PollingCounter[]? _counters;
+#endif
+
+    [NonEvent]
+    public void ResetCounters()
+    {
+        Volatile.Write(ref _totalLocalCacheHit, 0);
+        Volatile.Write(ref _totalLocalCacheMiss, 0);
+        Volatile.Write(ref _totalDistributedCacheHit, 0);
+        Volatile.Write(ref _totalDistributedCacheMiss, 0);
+        Volatile.Write(ref _totalBackendExecute, 0);
+        Volatile.Write(ref _currentBackendExecute, 0);
+        Volatile.Write(ref _currentDistributedFetch, 0);
+        Volatile.Write(ref _totalLocalCacheWrite, 0);
+        Volatile.Write(ref _totalDistributedCacheWrite, 0);
+        Volatile.Write(ref _totalStampedeJoin, 0);
+    }
+
+    [Event(EventIdLocalCacheHit, Level = EventLevel.Verbose)]
+    public void LocalCacheHit()
+    {
+        DebugAssertEnabled();
+        _ = Interlocked.Increment(ref _totalLocalCacheHit);
+        WriteEvent(EventIdLocalCacheHit);
+    }
+
+    [Event(EventIdLocalCacheMiss, Level = EventLevel.Verbose)]
+    public void LocalCacheMiss()
+    {
+        DebugAssertEnabled();
+        _ = Interlocked.Increment(ref _totalLocalCacheMiss);
+        WriteEvent(EventIdLocalCacheMiss);
+    }
+
+    [Event(EventIdDistributedCacheGet, Level = EventLevel.Verbose)]
+    public void DistributedCacheGet()
+    {
+        // should be followed by DistributedCacheHit, DistributedCacheMiss or DistributedCacheFailed
+        DebugAssertEnabled();
+        _ = Interlocked.Increment(ref _currentDistributedFetch);
+        WriteEvent(EventIdDistributedCacheGet);
+    }
+
+    [Event(EventIdDistributedCacheHit, Level = EventLevel.Verbose)]
+    public void DistributedCacheHit()
+    {
+        DebugAssertEnabled();
+        _ = Interlocked.Increment(ref _totalDistributedCacheHit);
```

Comment:
I gather it doesn't matter that these two counts aren't updated atomically (i.e. the increment might be observed without the decrement)?

Reply:
added

```
        // note: not concerned about off-by-one here, i.e. don't panic
        // about these two being atomic ref each-other - just the overall shape
```